Validate existing ESPHome config entry before DHCP based update

DanielBaulig opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->



## Proposed change
When the DHCP component notifies the ESPHome integration about a changed or updated IP address, we should validate the old entry first. If it still works, we will skip updating to the new entry. This ensures that the config entry doesn't get overridden in situations where it's configured in a way in which local IP address updates are irrelevant, like via a domain name or behind a NAT.
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: See https://community.home-assistant.io/t/refining-automatic-config-entry-updating-via-device-discovery/731881/6
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

DanielBaulig committed:
[esphome] Validate existing config entry before dhcp based update

Check if the existing config entry is still working before updating it with the DHCP notification.

DanielBaulig committed:
Fix lints

DanielBaulig committed:
More lint fixes

DanielBaulig committed:
Moar lint fixes

DanielBaulig committed:
Fix tests

bdraco commented on homeassistant/components/esphome/config_flow.py:
```diff
@@ -288,14 +288,56 @@ async def async_step_mqtt(
         # for configured devices.
+        assert entry is not None
+
+        # Check if the existing entry is still valid
+        error = await self.try_connect(entry)
```
We should do something like https://github.com/home-assistant/core/blob/04101b044b2618989ab35753219b62d66c60440a/homeassistant/components/powerwall/config_flow.py#L104 so we don't probe the device if config entry is setup and communicating

DanielBaulig replied:
Will do.

DanielBaulig replied:
So I looked into this. There's two problems I'm running into: 

1) I can't seem to identify any state that is similar to powerwall's `async_last_update_was_successful` state in the esphome integration or aioesphomeapi. If you have a recommendation / pointer, please let me know.
2) I added a guard on the `config_entry.state` like in Powerwall, to only allow probing when the config_entry is in retry or failure state. However, the `test_connection_aborted_wrong_device` test in test_manager.py started failing due to that change. If I understand the test correctly, then this test makes sure that the config entry gets changed if it connects to the configured node, but finds a node with a different name there. I'll need to make sure that the probing logic correctly fails for this test. I had `connect` throw an `APIConnectionError` so far, but I don't think that is the expected behavior. Instead, the connection should succeed, but the DeviceInfo name should be `"different"` (vs the expected `"test"`). That said, the actual issue I'm running into is with guarding with `config_entry.state`. At least during test execution `config_entry.state` does not get changed to any of the failure states upon the wrong DeviceInfo being reported. I realized that I am not sure what the correct expectations here are? Maybe I am misunderstanding how this is supposed to work, but should the config_entry be set to a SETUP_ERROR or SETUP_RETRY state upon connecting to a wrong device? If not, then I cannot guard against `config_entry.state` if other parts of the integration rely on the DHCP update to be correctly processed here.

DanielBaulig replied:
I think I just found `entry_data.available`, which might solve for point 1) 
For point 2) I would still appreciate some clarification though.

bdraco replied:
For #2 ESPHome always sets up an never enters a retry state because it never raises `ConfigEntryNotReady`. This is mostly for historical reasons since before all the retry logic was built-in. The only time is would enter setup error is if there was an unhandled exception in during setup. We should not plan for this case as it would be a bug that should be fixed

DanielBaulig replied:
Thanks. I'm not 100% sure if I understood what you're suggesting to do in the context of this PR. You're saying I should ignore `config_entry.state` for now and only depend on `entry_data.available`, since it would cause a bug otherwise? I can add a comment in the code that `config_entry.state` should be added to the check once ESPHome's config entry state management was modernized. Does that make sense?

bdraco commented on homeassistant/components/esphome/config_flow.py:
```diff
@@ -288,14 +288,56 @@ async def async_step_mqtt(
 
+    async def try_connect(self, entry: ConfigEntry) -> str | None:
+        """Try connecting to a device and return any errors."""
+        host = entry.data[CONF_HOST]
+        port = entry.data[CONF_PORT]
+        password = entry.data[CONF_PASSWORD]
+        psk = entry.data.get(CONF_NOISE_PSK, "")
+
+        zeroconf_instance = await zeroconf.async_get_instance(self.hass)
+
+        cli = APIClient(
+            host,
+            port,
+            password,
+            zeroconf_instance=zeroconf_instance,
+            noise_psk=psk,
+        )
+
+        try:
+            await cli.connect()
+        except APIConnectionError:
+            return "connection_error"
+        finally:
+            await cli.disconnect(force=True)
+
```
This code is substantially similar to `fetch_device_info`. It would be good to split `fetch_device_info` into two functions so you can reuse part of it here.

DanielBaulig replied:
Yeah, I thought about it. But the similarity is really only in three minor aspects:

1) acquiring zeroconf instance
2) creating APIClient
3) calling connect on APIClient instance

`fetch_device_info` (as well as `try_login`) differ from `try_connect` in where they obtain their data from as well as how they respond to various exceptions and failures. I can extract the three above mentioned parts into their own method, but it's not going to be as significant as one might expect at first.

DanielBaulig committed:
Extract APClient creation and require config_entry failure state

Note: test_connection_aborted_wrong_device in test_manager.py currently fails.
See comments on pull request.

frenck commented:
This is odd logic IMHO. We use this logic in many more places, and making exceptions here seems, therefore, wrong?

DanielBaulig commented:
@frenck Thanks for the feedback. 

ESPHome wouldn't be the first or only integration to verify it's old connection before adopting the new connection information. There are some examples of that behavior in other integrations pointed out in this thread: https://community.home-assistant.io/t/refining-automatic-config-entry-updating-via-device-discovery/
There are also at least several more integrations that I came across looking for ConfigEntryState.SETUP_ERROR across the codebase yesterday. They all follow this pattern of: only adopt the new DHCP information if the config entry is in an error state and/or a working connection cannot be established.

I agree with your notion though and my original instinct was to fix this on the platform level (i.e. in Home Assistant core), too but that didn't seem feasible after some additional discussion and exploration. 

Home Assistant or the DHCP component cannot really tell or verify if an integration can still connect to it's device using the previous configuration. Implementation details matter a lot here and only the integration can really make this call.

Therefor my original intention was to use a heuristic based on the previous configuration. I.e. if the configured host is a globally routable IP address or FQDN, it seems unlikely that updating the configuration with a local IP address is the desired behavior.

The feedback I got was (both in the community forum as well as on Discord) to instead have the integration check if the previous configuration is still functional, like these other integrations do. 

I also realized that my heurisitcs approach wouldn't have been possible. I don't think the DHCP component knows or can know what the currently configured IP address / host for the integration is. While convention seems to be to store that information in `entry[CONF_HOST]`, I don't think that's *necessarily* the case. So ultimately the integration would have to be the one to make that call for a heuristics approach as well.

frenck commented:
> ESPHome wouldn't be the first or only integration to verify it's old connection before adopting the new connection information it receives from device discovery. 

I honestly don't think that logic is correct. Just because it's connected right now doesn't mean it is the right path it takes. This sentence contains (just as the existing solution) an assumption.

> There are some examples of that behavior in other integrations

Understood, I suspected it does, which is also why I commented on this PR.

>  They all follow this pattern of: only adopt the new DHCP information if the config entry is in an error state and/or a working connection cannot be established.

Correct, as we assume a normal, flat network for Home Assistant. Any deviation of it, would be an advanced use case, which we will not take into account on an application level.

> Home Assistant or the DHCP component cannot really tell or verify if an integration can still connect to it's device using the previous configuration.

IMHO that doesn't matter, there might be more routes to an device. This is an assumption as well.

To conclude, I don't like this approach added to this PR, and I most certainly don't want to see every integration implementing this logic over and over again.

Instead, if this causes an issue for you, I think you should adjust your network to not leak DHCP information across you different networks OR just disable the DHCP integration.

../Frenck

DanielBaulig commented:
I appreciate your feedback and you sharing your input. I will address various points individually, if you don't mind.

> I honestly don't think that logic is correct. Just because it's connected right now doesn't mean it is the right path it takes. This sentence contains (just as the existing solution) an assumption. I rather see an assumption across or system that is consistent, than making exceptions on a single integration (in this case, ESPHome).

I suspect I might not (fully) understand the purpose of the DHCP component and the device discovery mechanism in relation to updating existing config entries then. You say that "just because it's connected right now doesn't mean it is the right path it takes". But isn't that *exactly* why one might want to utilize device discovery to reconfigure existing config entries? To be able to "automagically" recover from connectivity issues caused by network changes, like new IP address assignments? If the existing configuration continues to work, why update it? See below for some reasons why one might *not* want to update it in those scenarios.

> Understood, I suspected it does, which is also why I commented on this PR.

Make sense. Thanks!

> Correct, as we assume a normal, flat network for Home Assistant. Any deviation of it, would be an advanced use case, which we will not take into account on an application level.

I think it makes sense to assume a flat, simple network as a baseline. Obviously a vast majority of environments HA will find itself in will be exactly that. And as such it makes sense that they may not be a *priority* for the platform. But why would Home Assistant not *care* about such advanced use-cases or even outright reject them / "not take [them] into account"? There are tradeoffs of course. And I there are concerns, I would love to take a closer look at them and discuss those.

> IMHO that doesn't matter, there might be more routes to an device. This is an assumption as well.

I 100% agree. This is the reason I started this discussion and pull request. There *can* be more than one route to a device. And if a user has configured a specific route, by the virtue of using a specific IP address or hostname, and that route is still working upon the DHCP component "seeing" that same device on any other network, then one ought to not update that configuration if one doesn't want to risk to break the connection right now or at some point in the future. 

Of course, if the config entry was previously auto configured via device discovery, you are probably right: not updating the config entry wouldn't make any sense. 
But if the user has manually and explicitly set a specific IP address or hostname it becomes more murky. Especially if the hostname or IP address are not local (but there are purely local scenarios as well). I am not saying it never makes sense to update the configuration even in those cases, but it becomes less straight forward.

> I don't like this approach added to this PR, and I most certainly don't want to see every integration implementing this logic over and over again. This would be highly inefficient and duplicate code, that add complexity, mostly to support a use case we shouldn't be caring about on an application level.

I hear you. I think you are raising some valid concerns. But if so many integrations are going out of their way to implement this logic or logic similar to this, then there must be a reason for that. All these integrations and/or their users *must* be running into problems that get fixed by this logic. Why would they go through the effort of implementing it otherwise? 

Would a less fragmented solution be preferable? Probably. Does this add complexity? Certainly. Does it address a real problem that real people have in the real world? It appears so.

> Instead, if this causes an issue for you, I think you should adjust your network to not leak DHCP information across your different networks OR just disable the DHCP integration.

I appreciate your suggestion on how to fix my particular issue. In my particular case it was my router integration "leaking" that information via device trackers. I was able to work around this particular issue by disabling this feature in my router integration. I can go into more detail here, but I am not sure that is really useful.

The main reason I am looking for a more systemic solution rather than just a fix on my end is that I am building out a service with the goal of making it easy for people to deploy ESPHome nodes outside their local network while still integrating them into their Home Assistant instance. Most people seem to setup those nodes on their local network first using the remote connectivity profile and then move them to the remote location once they have established it's working. But when the DHCP component sees the node on the local network, the config entry will get updated with the local connectivity information. Once people move their node to the remote location, Home Assistant cannot connect to it anymore. I believe users rolling their own ESPHome/WireGuard setup might be already running into this problem, too.

I can provide some additional entirely local examples where current behavior will cause issues for people with non-simple, non-flat network setups (like my personal router/device tracker example from above), if that is something you might be interested in. I understand that those setups are not a priority for Home Assistant as a platform.

As of today, remote ESPHome nodes might not be an incredible common use-case. But I believe part of that is because it's too hard to setup and configure. And I know that I can make this more approachable and easier. Fixing this particular issue is going to go a long way with that and open up interesting new use-cases for Home Assistant and ESPHome that were previously impossible or at least prohibitively hard to setup.

I hope this makes sense and hope I can get your support *in principle*. Happy to be flexible and accommodating when it comes to the particulars.  Thanks for all the amazing work you and everyone else at Nabu Casa have been putting into Home Assistant <3

bdraco commented:
This one is a difficult call as we want to avoid supporting every complex network configuration, as that tends to be a never-ending rabbit hole. For the integrations we are currently checking and not updating from discovery, the driving factor is that those devices have multiple network interfaces (either connected to various VLANS or wifi+ethernet on the same). ESPHome devices could fall into that category, depending on the hardware.

I've also tried many times before to come up with a solution for this issue every time it comes up. Still, there isn't something one-size-fits-all that will work for the core. We usually end up with a per-integration solution each time since the discovery data and availability check often differ between different integrations. I'd prefer this was handled by core, and integrations didn't have to implement something like this. However, I'm not sure it's possible to accomplish that, given how non-standardized discovery implementations/data structures/methods are between devices.

> Once people move their node to the remote location, Home Assistant cannot connect to it anymore. I believe users rolling their own ESPHome/WireGuard setup might be already running into this problem, too.

That probably is a good justification for implementing reconfigure in ESPHome so its easier to change the IP.

DanielBaulig commented:
Given that the ESPHome team cares for the VPN/remote case and there is no solution for this problem in core, any chance we can revisit this PR or discuss some alternative approaches that can solve for the VPN/remote issue caused by automatic device discovery based reconfiguration?

edenhaus commented:
We discussed this topic internally and decided to close this PR, as we agreed with @bdraco and @frenck's comments. We don't want custom handling in each integration for edge cases like an advanced network, where you have DHCP information leakage.
 
> > Once people move their node to the remote location, Home Assistant cannot connect to it anymore. I believe users rolling their own ESPHome/WireGuard setup might be already running into this problem, too.
> 
> That probably is a good justification for implementing reconfigure in ESPHome so its easier to change the IP.

As suggested we should implement reconfigure in ESPHome, so a user can easily change the IP.

> Instead, if this causes an issue for you, I think you should adjust your network to not leak DHCP information across your different networks OR just disable the DHCP integration.

As already mentioned, each user has the possibility to disable DHCP discovery completely or fix his network. In my opinion, your issue is a not-correctly configured network; otherwise, the assigned IP address in the DHCP response would be valid and reachable.
